### PR TITLE
[ZK filter] Add per opcode request/response bytes counters

### PR DESCRIPTION
**Pull request by Winbobob**
Commit Message: [ZK filter] Add per opcode request/response byte counters
Additional Description: Currently, the ZK filter only reports total request/response bytes. This PR adds the feature to report  per opcode request/response bytes counters. These new metrics will be controlled by `enable_per_opcode_request_response_bytes`. Its default value is false, which means these metrics will not be emitted by default.

With the per opcode request/response bytes counters, plus existing per opcode request/response counters, we are able to get the average request/response size of each opcode. This gives us more observability of ZK.

Risk Level: low
Testing: unit tests
Docs Changes: This PR updates the [ZK filter doc](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/zookeeper_proxy_filter).
Release Notes: See changelogs/current.yaml
[API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md): Add `enable_per_opcode_request_bytes` and `enable_per_opcode_response_bytes`  to ZK proto.

### CODE REVIEW

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (adisuissa) [inline source/extensions/filters/network/zookeeper_proxy/filter.cc:271]: nit: move next to its usage place (inside the `if` scope).

File: api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:64]: Do you think there will be a benefit to emit just the request or the response metrics?
If so, I suggest splitting it to 2 fields.

File: changelogs/current.yaml
Comment (adisuissa) [inline changelogs/current.yaml:86]: Please add a reference to the field

File: docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst
Comment (adisuissa) [inline docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:44]: I think these new fields also need the `enable_per_opcode_request_response_bytes_metrics` field to be set. If so, please update this doc to clarify which stats are enabled by which flag.

File: docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst
Comment (adisuissa) [inline docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:126]: High-level comment - stats are not free, and it seems that many are added here, which may impact performance. Have you considered other mechanisms to achieve your goal?

File: source/extensions/filters/network/zookeeper_proxy/filter.h
Comment (adisuissa) [inline source/extensions/filters/network/zookeeper_proxy/filter.h:279]: I think these can be private

File: source/extensions/filters/network/zookeeper_proxy/filter.h
Comment (adisuissa) [inline source/extensions/filters/network/zookeeper_proxy/filter.h:305]: Same as above.

**Review by adisuissa — COMMENTED** [review]
Thanks, left one high-level comment, and mostly minor things.
Please also add an integration test ensuring that the feature works.

File: docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst
Comment (Winbobob) [reply docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:126]: Although ZooKeeper 3.6.0 introduced [a new metric system](https://zookeeper.apache.org/doc/current/zookeeperMonitor.html#:~:text=Four%20letter%20words-,New%20Metrics%20System,%2C%20requestProcessor%2C%20and%20so%20forth.), it does not provide the per opcode request/response bytes metrics, which could be helpful to understand the payload size of different opcodes during debugging. This PR also introduces switch(s) to enable/disable these metrics. Then users could stop emitting these metrics if not needed.

Comment (Winbobob) [reply docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:132]: Yeah, you are right! Let me refactor the code to initial new metrics in filter config.

**Comment by Winbobob** [comment]
Hi @JuniorHsu, please review the PR if you have a chance, thank you!

File: api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto
Comment (JuniorHsu) [inline api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:63]:
```suggestion
  // Whether to emit per opcode request bytes metrics. If not set, it defaults to false.
```

File: api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto
Comment (JuniorHsu) [inline api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:66]:
```suggestion
  // Whether to emit per opcode response bytes metrics. If not set, it defaults to false.
```

File: docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst
Comment (JuniorHsu) [inline docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:38]: this is subtle. explicit say it's per method metrics

File: source/extensions/filters/network/zookeeper_proxy/decoder.h
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/decoder.h:159]: comment the return value and what does absl::nullopt mean

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/filter.cc:248]: is it possible that the opcode is not in the map? if it's impoosible, assert and log error

**Review by JuniorHsu — CHANGES_REQUESTED** [review]
lgtm modulo some nits and questions

File: source/extensions/filters/network/zookeeper_proxy/decoder.h
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/decoder.h:83]: hard to understand the difference between `onDefaultRequestBytes` and `onRequestBytes`
either add some comment or changing the latter signature?

File: source/extensions/filters/network/zookeeper_proxy/decoder.h
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/decoder.h:160]: did we test WATCH_XID?

File: source/extensions/filters/network/zookeeper_proxy/decoder.h
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/decoder.h:159]:
```suggestion
  // decodeOnData and decodeOnWrite return ZooKeeper opcode or absl::nullopt.
```

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/filter.cc:250]: probably 
```
ASSERT(config_->op_code_map_.contains(opcode));
config_->op_code_map_[opcode].rq_bytes_counter_add(bytes);
```
wdyt?

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/filter.cc:258]: let's do if else instead of switch

**Review by JuniorHsu — CHANGES_REQUESTED** [review]

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc:250]: Thanks, it is much more clean, will make the change accordingly.

File: source/extensions/filters/network/zookeeper_proxy/decoder.h
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/decoder.h:83]: I removed `onDefaultRequestBytes` and merged its code into `onRequestBytes`.

File: source/extensions/filters/network/zookeeper_proxy/decoder.h
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/decoder.h:160]: Yeah, we tested here: https://github.com/envoyproxy/envoy/pull/29081/files#diff-a0f6243d3325c5bb6bbda5bb0e9c6a47723f25d95934e2edf05a98dc978f8f4fR1373-R1390

**Review by JuniorHsu — APPROVED** [review]
thanks

**Comment by JuniorHsu** [comment]
seems Adi is vacationing per his slack status 

/unassign @adisuissa
/assign-from @envoyproxy/api-shepherds

**Comment by JuniorHsu** [comment]
try again
/assign-from https://github.com/orgs/envoyproxy/teams/api-shepherds
/unassign @adisuissa

**Comment by JuniorHsu** [comment]
@mattklein123 no luck to pick a random api-shepherd. Could you help on api parts? thanks

**Comment by wbpcode** [comment]
ping @mattklein123 for API review.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc
Comment (adisuissa) [inline source/extensions/filters/network/zookeeper_proxy/decoder.cc:176]: I see that line 171 above throws an exception if the opcode is unknown. IIUC this method can be called on the data-path (not on the config). If so, please change the exception throwing to returning a nullopt or a value that signifies that an error occurred.

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (adisuissa) [inline source/extensions/filters/network/zookeeper_proxy/filter.cc:254]: What ensures that the `opcode` is always in the `op_code_map_`?
If it is certain, then please add a comment explaining why this invariant holds. Otherwise, add a if-else closure to handle the case where the op-code is unknown.

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (adisuissa) [inline source/extensions/filters/network/zookeeper_proxy/filter.cc:269]: Same as above.

File: docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst
Comment (adisuissa) [reply docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:134]: FWIW, I don't think the refactor addressed the issue of Envoy counters being initialized for every instantiation of the filter.

**Review by adisuissa — COMMENTED** [review]
Apologies, I've been OOO.
Left a few comments on the code.
/lgtm api

File: docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst
Comment (Winbobob) [reply docs/root/configuration/listeners/network_filters/zookeeper_proxy_filter.rst:134]: My understanding is that these new counters are added to the filter config, which will be shared by all filters.
Not sure whether you have some examples that I could refer to, thanks

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc:176]: This excaption will be caught later: https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/zookeeper_proxy/decoder.cc#L618-L632

Maybe we could keep it as is?

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc:269]: This opcode comes from decoder.cc "decodeOnWrite" function. It should match the opcode decoded by the "decodeOnData" function.

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (adisuissa) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc:254]: Correct me if I'm wrong, but `deocdeOnData` is being called on the data-path (processing a user's request/response), and not only on the config-path (processing a config update).
Envoy does not allow exceptions on the data-path - if this throws an exception, that is not being caught, then the proxy will segfault.

I don't understand exactly what's going on here, but there might also be a case where a user will send a request containing a wrong op-code, and that will cause an exception to be thrown. If this is the case, then it could be an attack vector on the proxy.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc
Comment (adisuissa) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc:176]: Due to some security consideration, Envoy does not allow [exceptions on the data-path](https://github.com/envoyproxy/envoy/blob/a19966b92bc3afcdd68ebeffe53c4b6848733e79/STYLE.md?plain=1#L26). I suggest changing this behavior (in a different PR), to comply with the rest of Envoy

**Review by adisuissa — DISMISSED** [review]
LGTM for the changes related to this PR.
I suggest following up with updates to the code to make it more secure.

**Comment by Winbobob** [comment]
> LGTM for the changes related to this PR.
> 
> I suggest following up with updates to the code to make it more secure.

Thank you for suggestions! Will do it in my next PR.

**Comment by adisuissa** [comment]
Please resolve the conflict.

/assign-from @envoyproxy/senior-maintainers

**Comment by adisuissa** [comment]
Please merge main again.
/unassign @phlax 
@RyanTheOptimist can you do the senior reviewer pass on this PR?
/assign @RyanTheOptimist

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (RyanTheOptimist) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc:254]: Is this going to be addressed in the follow up PR as well?

**Review by RyanTheOptimist — APPROVED** [review]
Assuming this unresolved comment will be addressed in a follow up PR, this LGTM.

File: source/extensions/filters/network/zookeeper_proxy/filter.cc
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc:254]: Yes, I will try to fix it in the next PR.

**Review by RyanTheOptimist — APPROVED** [review]

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc:176]: Hi @adisuissa, here is the PR to remove EnvoyException from the ZK proxy filter data plane: https://github.com/envoyproxy/envoy/pull/30438

Could you elaborate why keeping EnvoyException in data plane will have the security issues? Thanks!

Comment (adisuissa) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc:176]: There's prior work on removing exceptions from the data-plane, e.g., #10878.

**Pull request merged**
